### DISCUSSION TITLE
Access `NextRequest` in middleware?

### DISCUSSION THREAD

**Discussion by sleeyax** [next-safe-action/next-safe-action#270] [Q&A] [answered]
Is it possible to access the `NextRequest` object in middleware functions? 

I'm looking for a way to integrate `clerk` with `next-safe-action`. With TRPC you can simply [access the request object](https://clerk.com/docs/references/nextjs/trpc#create-clerk-authentication-context) as follows:

```ts
import * as trpc from '@trpc/server'
import * as trpcNext from '@trpc/server/adapters/next'
import { getAuth } from '@clerk/nextjs/server'

export const createContext = async (opts: trpcNext.CreateNextContextOptions) => {
  return { auth: getAuth(opts.req) }
}

export type Context = trpc.inferAsyncReturnType<typeof createContext>
```

How would I do this with `next-safe-action` (without using `cookies()`)?

**Comment by TheEdoRan** [comment, marked as answer]
Hi, currently no way to do it, unfortunately. That's not a limitation of this library, it's a limitation of React Server Actions, they don't expose the underlying POST request, you can only access `headers()` and `cookies()` via Next.js exported functions. You'd have to use route handlers if you want to access the request object.

**Reply by sleeyax** [reply]
Ah I see. That's too bad but I'll figure a way to work around it anyways. Thanks you for responding so quickly!
